servicetalk-grpc-protoc/src/main/java/io/servicetalk/grpc/protoc/Generator.java

```diff
@@ -248,13 +248,15 @@ private State(ServiceDescriptorProto serviceProto, GenerationContext context, St
     private final Map<String, ClassName> messageTypesMap;
     private final ServiceCommentsMap serviceCommentsMap;
     private final boolean printJavaDocs;
+    private final boolean skipDeprecated;
 
     Generator(final GenerationContext context, final Map<String, ClassName> messageTypesMap,
-              final boolean printJavaDocs, SourceCodeInfo sourceCodeInfo) {
+              final boolean printJavaDocs, final boolean skipDeprecated, SourceCodeInfo sourceCodeInfo) {
```

Comment:
Overall approach lgtm, but I added new option to our protoc-options example, run `./gradlew :servicetalk-examples:grpc:servicetalk-examples-grpc-protoc-options:build`, and this is what I found: 

1. Client interfaces are empty now:

before:
```java
    public interface GreeterClient extends GrpcClient<BlockingGreeterClient> {
        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        Single<HelloReply> sayHello(HelloRequest request);

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
         * @param metadata the metadata associated with this client call.
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        @Deprecated
        default Single<HelloReply> sayHello(SayHelloMetadata metadata, HelloRequest request) {
            return Single.failed(new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily."));
        }

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param metadata the metadata associated with this client call.
         * @param request the request to send to the server.
         * @return a {@link Single} which completes when the response is received from the server.
         */
        default Single<HelloReply> sayHello(GrpcClientMetadata metadata, HelloRequest request) {
            return sayHello(new SayHelloMetadata(metadata), request);
        }

        @Override
        default BlockingGreeterClient asBlockingClient() {
            return new GreeterClientToBlockingGreeterClient(this);
        }
    }

    public interface BlockingGreeterClient extends BlockingGrpcClient<GreeterClient> {
        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        HelloReply sayHello(HelloRequest request) throws Exception;

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @deprecated Use {@link #sayHello(GrpcClientMetadata,HelloRequest)}.
         * @param metadata the metadata associated with this client call.
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        @Deprecated
        default HelloReply sayHello(SayHelloMetadata metadata, HelloRequest request) throws
                Exception {
            throw new UnsupportedOperationException("This method is not implemented by " + getClass() + ". Consider migrating to an alternative method or implement this method if it's required temporarily.");
        }

        /**
         * <pre>
         *  Sends a greeting
         * </pre>
         *
         * @param metadata the metadata associated with this client call.
         * @param request the request from the client.
         * @return the response from the server.
         * @throws Exception if an unexpected application error occurs.
         * @throws io.servicetalk.grpc.api.GrpcStatusException if an expected application exception occurs. Its contents will be serialized and propagated to the peer.
         */
        default HelloReply sayHello(GrpcClientMetadata metadata, HelloRequest request) throws
                Exception {
            return sayHello(new SayHelloMetadata(metadata), request);
        }
    }
```

after:
```java
    public interface GreeterClient extends GrpcClient<BlockingGreeterClient> {
        @Override
        default BlockingGreeterClient asBlockingClient() {
            return new GreeterClientToBlockingGreeterClient(this);
        }
    }

    public interface BlockingGreeterClient extends BlockingGrpcClient<GreeterClient> {
    }
```

Expectation: they should have 2 methods: `sayHello(HelloRequest)` and `sayHello(GrpcClientMetadata metadata, HelloRequest request)`. The last one should not have `default` implementation anymore, bcz that `default` was added for backward compatibility only.

2. There are still some fields and methods with `@Deprecated` annotation:

```java
    @FunctionalInterface
    public interface BlockingSayHelloRpc extends BlockingGrpcService {
        /**
         * @deprecated Use {@link #methodDescriptor}.
         */
        @Deprecated
        String PATH = "/helloworld.Greeter/SayHello";
```

```java
            @Override
            @Deprecated
            protected GreeterServiceFromRoutes newServiceFromRoutes(
                    final GrpcRoutes.AllGrpcRoutes routes) {
                return new GreeterServiceFromRoutes(routes);
            }
```

Consider following the same approach for tests. However, existing `helloworld.proto` used by `protoc-options` has only one endpoint (aggregated). Only for the purpose of local testing, consider also copying `test_service.proto` into `protoc-options`. That will codegen all API variations we have. Would be better to use for `skipDeprecated=false|true` diff comparison.

Reply:
Fixed in the latest diff.